File: MoreHighlight/src/main/java/xyz/wingio/plugins/morehighlight/AliuRule.java
package xyz.wingio.plugins.morehighlight;

import android.content.Context;

import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.Fragment;

import com.aliucord.PluginManager;
import com.aliucord.Utils;
import com.aliucord.utils.*;
import com.aliucord.entities.Plugin;

import com.discord.app.*;
import com.discord.simpleast.core.parser.ParseSpec;
import com.discord.simpleast.core.parser.Parser;
import com.discord.simpleast.core.node.Node;
import com.discord.simpleast.core.parser.Rule;
import com.discord.utilities.textprocessing.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import kotlin.Unit;

public final class AliuRule extends Rule<MessageRenderContext, ClickableNode<MessageRenderContext>,MessageParseState> {
    private Context context;
    private static FragmentManager cachedFragment = Utils.appActivity.getSupportFragmentManager();

    public AliuRule(Pattern pattern, Context context) {
        super(pattern);
        this.context = context;
    }

    @Override
    public ParseSpec<MessageRenderContext, MessageParseState> parse(Matcher matcher, Parser<MessageRenderContext, ? super ClickableNode<MessageRenderContext>, MessageParseState> parser, MessageParseState s) {
        final Plugin p = PluginManager.plugins.get(matcher.group(1));
        Context ctx = context;
        var textNode = (p != null && p.settingsTab != null) ? new ClickableNode(matcher.group(1), "", context) : new TextNode(matcher.group(1));
        if(textNode instanceof ClickableNode) ((ClickableNode) textNode).setOnClickListener(v -> {
            try{
                if (p.settingsTab.type == Plugin.SettingsTab.Type.PAGE && p.settingsTab.page != null) {
                    Fragment page = p.settingsTab.args != null
                            ? ReflectUtils.invokeConstructorWithArgs(p.settingsTab.page, p.settingsTab.args)
                            : p.settingsTab.page.newInstance();
                    Utils.openPageWithProxy(ctx, page);
                } else if (p.settingsTab.type == Plugin.SettingsTab.Type.BOTTOM_SHEET && p.settingsTab.bottomSheet != null) {
                    AppBottomSheet sheet = p.settingsTab.args != null
                            ? ReflectUtils.invokeConstructorWithArgs(p.settingsTab.bottomSheet, p.settingsTab.args)
                            : p.settingsTab.bottomSheet.newInstance();

                    sheet.show(cachedFragment, p.getName() + "Settings");
                }
            } catch (Throwable e) {}
            return Unit.a;
        });
        return new ParseSpec<>(textNode, s);
    }
}

File: MoreHighlight/src/main/java/xyz/wingio/plugins/morehighlight/rule/AliuRule.kt
@file:Suppress("MISSING_DEPENDENCY_CLASS", "MISSING_DEPENDENCY_SUPERCLASS")

package xyz.wingio.plugins.morehighlight.rule

import android.content.Context
import androidx.fragment.app.FragmentManager
import com.aliucord.PluginManager
import com.aliucord.Utils
import com.aliucord.Utils.openPageWithProxy
import com.aliucord.entities.Plugin
import com.aliucord.utils.ReflectUtils
import com.discord.simpleast.core.node.Node
import com.discord.simpleast.core.parser.ParseSpec
import com.discord.simpleast.core.parser.Parser
import com.discord.simpleast.core.parser.Rule
import com.discord.utilities.textprocessing.MessageParseState
import com.discord.utilities.textprocessing.MessageRenderContext
import xyz.wingio.plugins.morehighlight.node.ClickableNode
import xyz.wingio.plugins.morehighlight.node.TextNode
import java.util.regex.Matcher
import java.util.regex.Pattern

class AliuRule(
    private val context: Context
): Rule<MessageRenderContext, ClickableNode<MessageRenderContext?>, MessageParseState>(
    Pattern.compile("^ac://([A-Za-z0-9$]+)")
) {

    override fun parse(
        matcher: Matcher,
        parser: Parser<MessageRenderContext?, in ClickableNode<MessageRenderContext?>?, MessageParseState?>,
        s: MessageParseState?
    ): ParseSpec<MessageRenderContext?, MessageParseState?> {
        val pluginSettings = PluginManager.plugins[matcher.group(1)]?.settingsTab
            ?: return ParseSpec<MessageRenderContext?, MessageParseState?>(TextNode(matcher.group(0)!!), s)

        val node: Node<MessageRenderContext> = ClickableNode(matcher.group(1)!!, context, {
            when (pluginSettings.type) {
                Plugin.SettingsTab.Type.PAGE -> {
                    val page = if (pluginSettings.args != null) ReflectUtils.invokeConstructorWithArgs(pluginSettings.page, *pluginSettings.args) else pluginSettings.page.newInstance()
                    openPageWithProxy(context, page)
                }

                Plugin.SettingsTab.Type.BOTTOM_SHEET -> {
                    val sheet = if (pluginSettings.args != null) ReflectUtils.invokeConstructorWithArgs(pluginSettings.bottomSheet, *pluginSettings.args) else pluginSettings.bottomSheet.newInstance()
                    sheet.show(fragmentManager, pluginSettings::class.simpleName)
                }

                else -> {}
            }
        }, {})

        return ParseSpec(node, s)
    }

    companion object {
        private val fragmentManager: FragmentManager = Utils.appActivity.supportFragmentManager
    }

}